ci-operator/step-registry/github/downstream-sync/github-downstream-sync-commands.sh

```diff
@@ -0,0 +1,112 @@
+    | b64url
+)
+JWT="$HEADER.$PAYLOAD.$SIGNATURE"
+INSTALLATION_ID=$(curl -s -H "Authorization: Bearer $JWT" -H "Accept: application/vnd.github+json" \
+    "https://api.github.com/repos/${DOWNSTREAM_REPO}/installation" | jq -r .id)
+GITHUB_TOKEN=$(curl -s -H "Authorization: Bearer $JWT" -H "Accept: application/vnd.github+json" -X POST \
+    "https://api.github.com/app/installations/${INSTALLATION_ID}/access_tokens" | jq -r .token)
+
+
+# get the repo
+WORKDIR="$(mktemp -d)"
+cd "$WORKDIR"
+git clone --single-branch --branch "${DEFAULT_BRANCH}" "https://github.com/${DOWNSTREAM_REPO}" repo
+cd repo
+git remote add upstream "https://github.com/${UPSTREAM_REPO}"
+git fetch upstream "${DEFAULT_BRANCH}"
+git fetch origin "${DEFAULT_BRANCH}"
+
+# check if a d/s merge PR is already open. This requires the PR titles to have the
+# below format. As long as this automation is the one creating the PRs it will stay
+# in this format. If something/someone else had created a d/s merge PR without this
+# format, then we will still get a new one created with this script
+LAST_PR_NUM=$(
+  curl -s -H "Authorization: token $GITHUB_TOKEN" \
+    "https://api.github.com/repos/${DOWNSTREAM_REPO}/pulls?state=all&base=${DEFAULT_BRANCH}&per_page=100&sort=created&direction=desc" |
+    jq -r '.[] | select(.title|test("DownStream Merge \\[[0-9]{2}-[0-9]{2}-[0-9]{4}\\]")) | .number' | head -n1
+)
+if [[ -n "$LAST_PR_NUM" ]]; then
+  MERGED=$(
+    curl -s -H "Authorization: token $GITHUB_TOKEN" \
+      "https://api.github.com/repos/${DOWNSTREAM_REPO}/pulls/${LAST_PR_NUM}" |
+    jq -r .merged
+  )
+fi
+
+# if our last d/s merge PR is still not merged, then we don't need a new one yet
+if [[ "$MERGED" != "true" ]]; then
+  echo "Last PR #${LAST_PR_NUM} not merged; exiting."
+  exit 0
+fi
+
+# to save on overhead we don't need to open a new d/s merge PR until we have enough commits to bring in
+# the env var MIN_COMMITS can be used to control that number
+NEW_COMMITS=$(git rev-list origin/${DEFAULT_BRANCH}..upstream/${DEFAULT_BRANCH} --count)
+echo "Found $NEW_COMMITS new commits upstream."
+(( NEW_COMMITS < MIN_COMMITS )) && { echo "Not enough commits in upstream to make it worth opening a new PR; exiting."; exit 0; }
+
+# if we made it this far, we can create the merge and push the PR
+BRANCH="d/s-merge-$(date +%m-%d-%Y)"
+git checkout -b "$BRANCH" origin/${DEFAULT_BRANCH}
+# if there happens to be a merge conflict we can still push it and create a PR for the humans to take a look at
+# and resolve. leave a /hold comment on it to make it obvious
+if ! git merge -X theirs "upstream/${DEFAULT_BRANCH}"; then
+  git add -A
+  git commit -m "Merge upstream/${DEFAULT_BRANCH} into ${DEFAULT_BRANCH} with conflicts ($(date +%m-%d-%Y))"
+  CONFLICT=true
+fi
+git remote set-url origin "https://x-access-token:${GITHUB_TOKEN}@github.com/${DOWNSTREAM_REPO}"
+git push origin "$BRANCH"
+
+PR_TITLE="NO-JIRA: DownStream Merge [$(date +%m-%d-%Y)]"
```

Comment:
how are humans notified/informed if theres a conflict besides the commit message? do we want to include in the title that theres an issue or?

Reply:
lower in this script you will see that there will be a /hold put on the PR with a note about the conflict.

It probably makes sense to add a very clear note in the PR title as well. I'll do that.